wagtail/images/models.py

```diff
@@ -787,7 +787,17 @@ def background_position_style(self):
 
     def img_tag(self, extra_attributes={}):
         attrs = self.attrs_dict.copy()
+
+        # Lazy async loading to improve initial page load performance
+        attrs.update(
+            {
+                "decoding": "async",
+                "loading": "lazy",
+            }
+        )
+
         attrs.update(extra_attributes)
+
```

Comment:
ok - I see your reasoning but for now I have flagged this as needs design decision just so another few core team members can chime in on this global opt in to all images.

Potential downsides
* Images may not be set with width/height correctly and delayed loading could cause some content to jump
* logos and core images (e.g. hero images) may not want to be delayed
* some sites may not want to introduce inconsistencies between older browsers that do not support this and newer ones that do

Just so it's clear, I think this is a good idea generally but want others to give their thoughts and chime in on us making this call for frontend sites.

Historically Wagtail has tried to be less opinionated about what goes on in the frontend sites from my understanding.

Maybe it is worth adding some 'upgrade considerations' in the 4.0 release md to see what changes developers can expect and how to stop these added attributes.

Reply:
Yep, knowing how developers can be resistant to changes that originally seemed universally positive, I'd want to at least be ready with instructions on how to revert to the previous behaviour (that doesn't require changing every `{% image %}` tag). I think this is a good thing to introduce as a new default, but it should be overrideable too.

Maybe it could be a setting in the `wagtail.images` app's appconfig (like we did elsewhere for `wagtail.users`: https://docs.wagtail.org/en/stable/extending/customising_group_views.html#custom-edit-create-forms)?

Reply:
Opt-out functionality sounds good to me. Having it opt-out just for the frontend is a little complex, as there's a lot of code reuse (definitely a good thing).

Using `appconfig` feels perhaps a little heavyweight to me - is that how we normally do this kind of config, or would a setting be better? `WAGTAIL_LAZY_IMAGE_TAG`, for example.

Reply:
Typically we'd use settings for this kind of thing, but I'm on a general push to steer us towards appconfigs more - I feel that settings have become a disorganised dumping ground for micromanaging details, and I think it's a good thing to make "I don't like this behaviour, how do I change it" feel more like subclassing the relevant app rather than looking through a laundry list of things that can be changed.

Having said that, this particular setting is pretty fundamental to what wagtail.images does, and less egregiously micromanagey than, say, "number of items per page shown in the chooser", so I could get on board with this being an ordinary setting. One advantage of making it a setting is that we could consider keeping the Wagtail-level default as no extra attributes, but add the lazy-loading attributes to the project template's settings file so that it becomes a default just for new projects.

Reply:
I like where you're heading Matt. 

Maybe something like `WAGTAILIMAGES_BASE_ATTRS` 

Then the project template comes with 

`WAGTAILIMAGES_BASE_ATTRS = { "decoding": "async", "loading": "lazy"}`

These get added to all images using the image tag but can also get overridden on a per tag basis. 

One downside is that it risks developers putting a global bad value like id (which should be unique) or alt (which breaks accessibility). Probably not a huge problem though. 

Also this gives a way for those using JS driven work arounds (ie11 support for example) some other attribute like `data-lazy='true'` .

Reply:
Another approach could be a middleware one?

This way we can add the middleware in Wagtail (for a similar goal of changing the base attributes) and provide instructions on how to override it for public URLs?

https://docs.djangoproject.com/en/4.0/topics/http/middleware/

Each image usage should still be able to opt in / out of various attributes via the image tag.

Note: It is very possible I misunderstand the capabilities of Django middleware here :) - I have written approximately one (stackoverflow copy/paste) middleware plugin in my life haha.

Reply:
I think a page under https://docs.wagtail.org/en/stable/advanced_topics/images/index.html would be fine, it's ok to have a small page in this section.

We have https://docs.wagtail.org/en/stable/advanced_topics/images/renditions.html but I feel like that is not really the most suitable.

Reply:
I ended up putting it in https://docs.wagtail.org/en/stable/topics/images.html, as that's where we already talk about how to add custom attributes to single image tags